SCAutolib/controller.py

```diff
@@ -0,0 +1,375 @@
+                user_.key = user_.card_dir.joinpath("private.key")
+                _gen_private_key(user_.key)
+
+            csr = user_.gen_csr()
+            if user_.cert is None:
+                user_.cert = user_.card_dir.joinpath(
+                    f"cert-{user_.username}.pem")
+
+            if isinstance(user_, user.IPAUser):
+                self.ipa_ca.request_cert(csr, user_.username, user_.cert)
+            else:
+                self.local_ca.request_cert(csr, user_.username, user_.cert)
+
+        user_.card.enroll()
+
+    def cleanup(self):
+        ...
+
+    def _validate_configuration(self, params: {} = None):
+        """
+        Validate schema of the configuration file. If some value doesn't present
+        in the config file, this value would be looked in the CLI parameters
+
+        :param params: CLI arguments
+        :return:
+        """
+        # FIXME: any schema requires all values to be in the config file, and
+        #  only IP address of IPA server is accepted from CLI arguments.
+        #  Add loading of the values from params dict
+        # IP regex
+        # Specify validation schema for CAs
+        schema_cas = Schema(And(
+            Use(dict),
+            # Check that CA section contains at least one and maximum
+            # two entries
+            lambda l: 1 <= len(l.keys()) <= 2,
+            {Optional("local_ca"): {"dir": Use(Path)},
+             Optional("ipa"): {
+                 'admin_passwd': Use(str),
+                 'root_passwd': Use(str),
+                 Optional('ip_addr', default=None): Use(str),
+                 'server_hostname': Use(str),
+                 'client_hostname': Use(str),
+                 'domain': Use(str),
+                 'realm': Use(str.upper)}}),
+            ignore_extra_keys=True)
+
+        # Specify validation schema for all users
+        schema_user = Schema({'name': Use(str),
+                              'passwd': Use(str),
+                              'pin': Use(str),
+                              Optional('card_dir', default=None): Use(Path),
+                              'card_type': Or("virtual", "real", "removinator"),
+                              Optional('cert', default=None): Use(Path),
+                              Optional('key', default=None): Use(Path),
+                              'local': Use(bool)})
+
+        # Specify general schema for whole config file
+        schema = Schema({"root_passwd": Use(str),
+                         "ca": schema_cas,
+                         "users": [schema_user]})
+
+        return schema.validate(self.lib_conf)
+
+    @staticmethod
+    def _general_steps_for_virtual_sc():
+        """
+        Prepare the system for virtual smart card. Preparation means to
+        configure pcscd service and opensc module to be able correctly working
+        with virtual smart card. Also, repository for installing virt_cacard
+        package is added in this method.
```

Comment:
Shouldn't the virt_cacard repo be added in setup_system before we try to install virt_cacard?